src/walfilter.c

```diff
@@ -390,6 +390,84 @@ pgmoneta_filter_operation_delete(int* file_count, struct walfile*** walfiles, st
    return 0;
 }
 
+/**
+ * Filter out records with specific XIDs from WAL files
+ *
+ * @param file_count Pointer to the number of WAL files
+ * @param walfiles Pointer to the array of WAL file pointers
+ * @param xids Array of XIDs to filter out
+ * @param xid_count Number of XIDs in the array
+ * @return 0 on success, non-zero on failure
+ */
+int
+pgmoneta_filter_xids(int* file_count, struct walfile*** walfiles, int* xids, int xid_count)
+{
+#define XID_IS_FILTERED(xid) \
+        ({ int found = 0; \
+           for (int k = 0; k < xid_count; k++) { \
+              if ((transaction_id) xids[k] == (xid)) { found = 1; break; } \
+           } found; })
+
+   int records_marked = 0;
+
+   if (xids == NULL || xid_count <= 0)
+   {
+      pgmoneta_log_info("No XIDs provided for filtering");
+      return 0;
```

Comment:
Remember - have one place for success for exit of a function, and one place for error exit